### PR TITLE
[DEV-1910][chore] refactor target_table and historical_feature_table controller

### PR DESCRIPTION
**Pull request by jevonyeoh**
## Description
Refactor the two controllers to move the common logic into a base class for reusability.

## Related Issue
https://featurebyte.atlassian.net/browse/DEV-1910

## Type of Change

Label this pull request to place it under the correct category in Release Notes:

|        **Pull Request Label**         | **Category in Release Notes** |
|:-------------------------------------:|:-----------------------------:|
|       `enhancement`, `feature`        |          🚀 Features          |
| `bug`, `refactoring`, `bugfix`, `fix` |    🔧 Fixes & Refactoring     |
|       `build`, `ci`, `testing`        |    📦 Build System & CI/CD    |
|              `breaking`               |      💥 Breaking Changes      |
|            `documentation`            |       📝 Documentation        |
|            `dependencies`             |    ⬆️ Dependencies updates    |

## Checklist

- [ ] I have read the [`CODE_OF_CONDUCT.md`](https://github.com/featurebyte/featurebyte/blob/main/CODE_OF_CONDUCT.md) and [`CONTRIBUTING.md`](https://github.com/featurebyte/featurebyte/blob/main/CONTRIBUTING.md) guides.
- [ ] I have written tests for the changes made.
- [ ] I have written docstrings in [NumpyDoc format](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard)
- [ ] I have labeled my Pull Request correctly

### CODE REVIEW

File: featurebyte/routes/historical_feature_table/controller.py [hunk 1]
```diff
@@ -49,108 +54,49 @@ def __init__(
...
-        _ = verbose
-        historical_feature_table = await self.service.get_document(document_id=document_id)
-        if historical_feature_table.observation_table_id is not None:
-            observation_table = await self.observation_table_service.get_document(
-                document_id=historical_feature_table.observation_table_id
-            )
-        else:
-            observation_table = None
+    async def get_additional_info_params(
+        self, document: MaterializedTableDocumentT
```
Comment (YS-L) [inline featurebyte/routes/historical_feature_table/controller.py:93]: Can `MaterializedTableDocumentT` be replaced `HistoricalFeatureTableModel` in the type annotation here? The type parameter `MaterializedTableDocumentT` is already tied to `HistoricalFeatureTableModel` based on the class definition:

```
class HistoricalFeatureTableController(
    FeatureOrTargetTableController[
        HistoricalFeatureTableModel,
        HistoricalFeatureTableService,
        HistoricalFeatureTableList,
        HistoricalFeatureTableInfo,
    ],
):
   ...
```

File: featurebyte/routes/common/feature_or_target_table.py [hunk 2]
```diff
@@ -0,0 +1,270 @@
...
+    serving_names_mapping: Optional[dict[str, str]] = None
+
+
+class FeatureOrTargetTableController(
+    BaseMaterializedTableController[
+        MaterializedTableDocumentT, MaterializedTableDocumentServiceT, PaginatedDocument
+    ],
+    Generic[
+        MaterializedTableDocumentT, MaterializedTableDocumentServiceT, PaginatedDocument, InfoTypeT
+    ],
```
Comment (YS-L) [inline featurebyte/routes/common/feature_or_target_table.py:87]: Curious why isn't `TableCreateT` one of the type parameters here?

File: featurebyte/routes/common/feature_or_target_table.py [hunk 3]
```diff
@@ -0,0 +1,270 @@
...
+)
+PayloadT = TypeVar(
+    "PayloadT",
+    TargetTableTaskPayload,
+    HistoricalFeatureTableTaskPayload,
+)
+TableCreateT = TypeVar(
+    "TableCreateT",
+    TargetTableCreate,
+    HistoricalFeatureTableCreate,
```
Comment (YS-L) [inline featurebyte/routes/common/feature_or_target_table.py:65]: It seems a bit strange that all the derived classes have to be enumerated when defining the base class here.

File: featurebyte/routes/historical_feature_table/controller.py (hunk 1 above)
Comment (jevonyeoh) [reply featurebyte/routes/historical_feature_table/controller.py:93]: hm yea, let me try again. i was getting some lint errors on a previous iteration.

File: featurebyte/routes/common/feature_or_target_table.py [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+
+
+class FeatureOrTargetTableController(
+    BaseMaterializedTableController[
+        MaterializedTableDocumentT, MaterializedTableDocumentServiceT, PaginatedDocument
+    ],
+    Generic[
+        MaterializedTableDocumentT,
+        MaterializedTableDocumentServiceT,
+        PaginatedDocument,
```
Comment (YS-L) [inline featurebyte/routes/common/feature_or_target_table.py:86]: I wonder if these three existing parameters need to be declared again after inheriting from `BaseMaterializedTableController` which is generic over these.

**Review by YS-L — APPROVED** [review]

File: featurebyte/routes/common/feature_or_target_table.py (hunk 4 above)
Comment (jevonyeoh) [reply featurebyte/routes/common/feature_or_target_table.py:86]: yea, it looks like all the generic types need to be re-declared as otherwise i get some errors in pycharm.
